### DISCUSSION TITLE
xDrip+ Complication not received readings NULL

### DISCUSSION THREAD

**Discussion by Cryopier** [NightscoutFoundation/xDrip#3293] [Bug reports]
I have a Samsung Watch 5 and a Samsung Galaxy S22 Ultra running the latest nightly build of xDrip+ and the complication reads NULL, no matter how I refresh both devices.

xDrip+'s Wear OS version was installed utilizing Wear Installer, based on the most recent Nightly build, as well.

I am sourcing glucose data from the Dexcom G7 sensor.  Readings are displaying properly on the Samsung Galaxy S22 Ultra.

**Comment by Cryopier** [comment]
Are you ever going to address my issue?

**Reply by Navid200** [reply]
There are no developers currently addressing WearOS.
I personally have no experience with it.  So, I cannot do anything about it myself.

**Reply by Cryopier** [reply]
I'm not entirely sure it's a Wear OS issue.  Are we sure that it isn't.on xDrip+ Android's end?

**Reply by Navid200** [reply]
That is something that needs to be verified by a developer with WearOS experience.
It is quite possible that something needs to be updated in xDrip.  But, to verify and test, we need someone who can verify that.

**Reply by cwpeterson** [reply]
I just switched to G7 yesterday and updated xDrip and I'm seeing the same issue on my Pixel watch. I never had an issue prior to having to scan the QR code to support the G7. I really rely on this on my watch to track things during my work day. Whomever worked on the original complication, is no further development being done at all?

**Reply by IsaacRoitman** [reply]
Same issue on Pixel Watch 2, values are displayed properly in the xDrip watch faces.  xDrip complications in other watch faces display no BG data, just NULL or NULL NULL.

Pixel 5 running version e5113d9-2024.03.10

**Reply by bsw-del** [reply]
hey @Navid200 , thanks for all the help, just curious if there is any progress on this, or if I should pursue the tasker route mentioned below? I'd prefer an updated complication, as it is already all setup, and the tasker option seems quite involved. @IsaacRoitman , did you have any success getting this sorted?

**Reply by Navid200** [reply]
I am not aware of anyone working on making any change to xDrip to address WearOS.

MiBand is the companion watch that is still currently supported by a developer.

**Reply by IsaacRoitman** [reply]
Unfortunately not yet, I've installed Tasker on my phone but have not had time to get things setup and seems involved.  Agreed, the complication would be ideal since it worked perfectly before and almost works now.  And we have the ability to input treatments with Wear OS.  

@Navid200, anyone we can escalate to in the community?

**Reply by bsw-del** [reply]
Just an update on my side. I did the tasker and autowear combo described below. Very noticeable battery improvement on my pixel watch. It used to die after about 16 hrs of use. I'm at 30% and 26 hrs right now.
Didn't realize that xdrip was so heavy on wear os until now.

**Comment by opasha** [comment]
I just thought I'd chime in to help. I have the Ticwatch Pro 5 (Wear OS 3.5) and just got the OnePlus Watch 2 (Wear OS 4.0) today. I also got the famous Null Null nonsense again after the updating the xDrip+ Nightly about 3 updates ago with a running Dexcom G7. The xDrip+ default Watchfaces all work fine as you all have noted. 

I use the Marine Commander Watchface on all my watches, and the bottom complication circle is my xDrip+ Complication. I believe I figured out the issue. Maybe it works for you maybe not. 

Steps to Fix:
1) Go into the Wear OS xDrip+ Preferences on the watches and uncheck "Refresh on Change"
2) Wait the 1 minute or until it refreshes from your phone to your next reading if it doesn't after the first minute 
3) If still nothing yet, go to the xDrip+ phone app's main screen and press the three little dots in the top right corner and select "Resend glucose to watch" and it should force it to refresh on the watch. 
4) You can leave it like this or go back in the Wear OS xDrip+ Preferences and re-check "Refresh on Change" again and it should be fully fixed. 

At least for me, this fixed the issue on both my watches and updates now as my blood sugar changes. 

Let me know if it works for you all. All the best!

Wear Installer 2 still works great for Android 13/14 and Wear OS 3.5/4.0 to push APKs to your Wear OS watch, but you need to actually first Pair your watch with the Pair Code and Port, then go back enter in the full IP Address and Port (this Pairing Port is different from the previous Pairing Port for pairing your watch - this is very important). Here's a good video from the developer: 

https://www.youtube.com/watch?v=abgN4jQqHb0

**Reply by cwpeterson** [reply]
I may take a look at this again but with the issues and my need to have it working, I removed it from my watch and pushed forward with another option that uses Tasker to send the updates to the watch. It saves me a ton of battery too vs using the xDrip complication which has known issues with battery drain. Another bonus is that I don't have to update the watch at all when I update xDrip.

**Reply by opasha** [reply]
That's understandable as xDrip+ in general has battery drain due to how often it needs to access data from your phone for your readings, but to be fair, I didn't notice it have that much drain recently with the complication compared to the xDrip+ native Watchface. 

Yes, updating (especially if you are on the nightly update channel like me) can be quite annoying, but it's pretty easy with Wear Installer 2. I do wish it would push/update automatically to the watches as the developer intended. It "pretends" to update and send the updates to the watches whenever you update the phone app within the phone app itself (which I'm sure you've seen on the many recent updates), but that never actually works. It actually keeps trying to infinitely update and fails every time until you actually uninstall and manually push the apk update through Wear Installer 2 to the watch.

**Reply by IsaacRoitman** [reply]
This workaround did not work for my Pixel Watch 2.  Interested in Tasker until there is a fix, having our BG values on the watch is a huge value.  Feel free to link me to instructions or guides on how to, thanks!

**Reply by cwpeterson** [reply]
This is the thread comment where I learned about it: https://github.com/NightscoutFoundation/xDrip/discussions/2026#discussioncomment-6555382

Here is a direct link to his github and instructions, it's been working great for me. I had to buy the AutoWear and AutoNotification apps. Not a lot of money though considering what you get: https://gist.github.com/oNSXo/0706201d3ac28a99e1855d772e3039eb?permalink_comment_id=4351085#gistcomment-4351085

**Reply by bsw-del** [reply]
Any updates on the complication updating? I just moved to G7 and lost my complication on Pixel Watch (first gen) FWIW, the Xdrip watchface works, but just the complication is broken.. 
I'm going to try tasker tonight if can't get sorted.

**Reply by cwpeterson** [reply]
Tasker method has been working great for me until I had a sensor failure. The complication just never updated once it failed. So I didn't realize that something was going on until my pump started giving audible alarms. But this is still better than nothing. But new sensor and it's back up and running. Would be nice to have the xDrip option back since it shows timeouts and no readings etc.

**Reply by oblus** [reply]
> I just thought I'd chime in to help. I have the Ticwatch Pro 5 (Wear OS 3.5) and just got the OnePlus Watch 2 (Wear OS 4.0) today. I also got the famous Null Null nonsense again after the updating the xDrip+ Nightly about 3 updates ago with a running Dexcom G7. The xDrip+ default Watchfaces all work fine as you all have noted.
> 
> I use the Marine Commander Watchface on all my watches, and the bottom complication circle is my xDrip+ Complication. I believe I figured out the issue. Maybe it works for you maybe not.
> 
> Steps to Fix:
> 
> 1. Go into the Wear OS xDrip+ Preferences on the watches and uncheck "Refresh on Change"
> 2. Wait the 1 minute or until it refreshes from your phone to your next reading if it doesn't after the first minute
> 3. If still nothing yet, go to the xDrip+ phone app's main screen and press the three little dots in the top right corner and select "Resend glucose to watch" and it should force it to refresh on the watch.
> 4. You can leave it like this or go back in the Wear OS xDrip+ Preferences and re-check "Refresh on Change" again and it should be fully fixed.
> 
> At least for me, this fixed the issue on both my watches and updates now as my blood sugar changes.
> 
> Let me know if it works for you all. All the best!
> 
> Wear Installer 2 still works great for Android 13/14 and Wear OS 3.5/4.0 to push APKs to your Wear OS watch, but you need to actually first Pair your watch with the Pair Code and Port, then go back enter in the full IP Address and Port (this Pairing Port is different from the previous Pairing Port for pairing your watch - this is very important). Here's a good video from the developer:
> 
> https://www.youtube.com/watch?v=abgN4jQqHb0

Hi. I also have OnePlus Watch 2 and I cant get BG on complication. My phone is S10e connected to Libre2EU sensor over BT (xdrip+oop2). Do You still have yours visible?

btw. Im tryin to make my watch work as collector , did You tried that?? I cant connect to sensor with xdrip wear app (even with oop2 installed on watch). I only can connect with juggluco wear app but I dont like TCP type of connection with juggluco in phone.

**Comment by Navid200** [comment]
Can anyone identify exactly which xDrip Nightly was the very first that caused this problem?

**Reply by colel83** [reply]
The last correctly functioning nightly for me was 12-23-2023. Any I have tried after that show null on my galaxy watch 6.

**Reply by Navid200** [reply]
Thanks for that.
The release after December 23, 2023 is January 2, 2024.

When I look at the delta between the two releases, I see nothing related to Wear.  
The only way to move forward I can think of is the following:

1- Please install the January 2, 2024 and make sure it fails.  Let me know if it does.
2- Only if you confirm that release fails, I will create a test xDrip from the previous release and add only one of the changes and ask you to test it.

We will need to work together until we can figure out exactly what change caused this problem.
Depending on what we find, we can decide what the next course of action may be.

But, this will only work if you are certain that you have identified two consecutive releases where one fails and the other does not.  So, please kindly confirm and let me know.

**Reply by colel83** [reply]
So I was wrong. Jan 2nd build still works. But Jan 5th and Jan 10th builds do not.

**Reply by Navid200** [reply]
OK, thanks

**Reply by Navid200** [reply]
@colel83 What is your hardware data source?  Will you be able to run a test without disturbing your main setup and diabetes control?

**Reply by colel83** [reply]
Dexcom G7. I'm hoping so.

**Reply by Navid200** [reply]
Do you have a spare phone that is Android 7 or newer?
Can you install xDrip on it and set it up to be the follower of your main xDrip.  
Then, you can set up WearOS on it to see if it works or not  with a test xDrip I will give you.  Regardless of what happens, your main phone will remain untouched.

**Reply by PhilTU** [reply]
I have a WearOS watch on order and will be able to help with this without compromising anything else.

**Reply by heygar** [reply]
Any news on this? Best I can tell you is it's still broken on the April 19 build.
I was thinking of taking a stab at fixing it myself, but I think the learning curve might be too much for me just now. 
I was also thinking I might take a stab at doing a direct to watch version. But, that was more of a semi-interesting experiment once I learned the code better.

**Reply by heygar** [reply]
Version from 06/28/2024 seems to have the Complication Issue resolved. It's been running on my watch for a couple of days now...

**Reply by oblus** [reply]
Not for me. Im on S10e, xdrip+oop2 with Libre2EU on BT and OnePlus Watch 2.
Xdrip watch face wotking fine, but complication says "old Xmin" after set calibration on phone and before was "null null". I have unchecked "Refresh on Change".

**Reply by colel83** [reply]
Sorry I dropped off the conversation but I had some family stuff I had to deal with. I updated to the July 17th nightly on both my phone and watch and the watch complication is receiving data correctly again! Thanks for your hard work!!

**Reply by minebe4** [reply]
I'm on Galaxy watch 4 and Galaxy 24U. I haven't been getting readings on complication for a long time. Still don't. Is one of my settings wrong? 

I started just using the stock watch face but frankly the watch face is not great and doesn't really fit on my watch. 

Taking the time now to dig into it but seeing it's working for folks and not for me I must be doing something wrong?
